sdk/java/pulumi/src/main/java/com/pulumi/core/internal/Urn.java

```diff
@@ -190,20 +192,29 @@ public static QualifiedTypeName parse(String qualifiedType) {
                     () -> format("expected qualified type, type part to be not empty, got: '%s'", type)
             );
             final Optional<String> parent;
-            if (qualifiedTypeParts.length == 2) {
-                var pt = qualifiedTypeParts[0];
+
+            // Handle parents. URNs can have nested parents which should be joined with a "$".
+            if (qualifiedTypeParts.length >= 2) {
+                // Remove the last qualifiedTypeParts element as it's the type.
+                String[] parentList = new String[qualifiedTypeParts.length - 1];
+                var pt = String.join(ParentSeparator, parentList);
```

Comment:
parentList is empty at this point?

Reply:
correct.

Reply:
So pt = ""

Reply:
It turned out that it was a list of length - 1 nulls so the join was nonNull and nonEmpty so it was going past the require, but the overall value was overwritten so the rest of the logic worked.

Fixed it.